components/form-control/script.js

```diff
@@ -251,6 +251,19 @@ class GamefaceFormControl extends HTMLElement {
             if (submitEvent.defaultPrevented) return;
         }
 
+        for (let element of this.formElements) {
+            if (!this.toCustomElement(element).willSerialize()) continue;
+            const validation = this.hasErrors(this.toCustomElement(element));
+            if (!validation.hasError) continue;
+
+            let errorMessage = '';
+            for (let errorType of validation.errors) {
+                errorMessage += errorMessages.get(errorType)(element);
+            }
+            this.showError(errorMessage, element);
+            return;
+        }
+
         this.currentSubmitButton = event.currentTarget;
         switch (this.method.toLowerCase()) {
             case formMethods.GET.toLowerCase():
```

Comment:
```suggestion
        const formElementsCache = this.formElements;
        for (let element of formElementsCache) {
            if (!this.toCustomElement(element).willSerialize()) continue;
            const validation = this.hasErrors(this.toCustomElement(element));
            if (!validation.hasError) continue;

            let errorMessage = '';
            for (let errorType of validation.errors) {
                errorMessage += errorMessages.get(errorType)(element);
            }
            this.showError(errorMessage, element);
            return;
        }

        this.currentSubmitButton = event.currentTarget;
        switch (this.method.toLowerCase()) {
            case formMethods.GET.toLowerCase():
```

and then in the switch case you can make

                `this.makeRequest(formMethods.GET, null, `${this.action}?${this.getFormSerializedData(formElementsCache)}`)`

and 

              `this.makeRequest(formMethods.POST, this.getFormSerializedData(formElementsCache), this.action)`

Like this you will get the form elements once and will make one traverse of the DOM tree when you do this.formElements.

This comment is related with the previous comment where getFormSerializedData is refactored

Reply:
Yes, it's a lot better like this - changed.